Revise documentation for Snowplow Inspector extension (#373)
snowplow/documentation: docs/tutorials/tutorial-a-guide-to-using-the-snowplow-inspector-extension/index.md

---
title: "A Guide to Using the Snowplow Inspector Extension"
date: "2021-06-09"
sidebar_position: 90
toc_max_heading_level: 4
---

At Snowplow we frequently use the [Snowplow Inspector chrome extension](https://chrome.google.com/webstore/detail/snowplow-inspector/maplkdomeamdlngconidoefjpogkmljm?hl=en) to check for event emission on a web page - and we also recommend that our clients do the same when in the process of validating their tracking code.
This extension is an [Open Source project](https://github.com/poplindata/chrome-snowplow-inspector) by [Poplin Data](https://poplindata.com/), one of our partners operating in Australia and New Zealand.
In 2023, Poplin Data [was acquired by Snowplow](https://snowplow.io/blog/snowplow-acquires-poplin-data/) and became Snowplow Australia.

> Once added to Chrome the extension can be opened by [opening Developer Tools](https://developer.chrome.com/docs/devtools/open/) (usually <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>I</kbd> or on Mac <kbd>Cmd</kbd>+<kbd>Option</kbd>+<kbd>I</kbd>), where it has its own tab named ‘Snowplow’ (look for the tab adjacent to ‘Elements’, ‘Console’, ‘Network’, etc. - it may be necessary to expand the list of tabs using the chevrons button (»)). 

Reviewer: ```suggestion
Once you add the extension to Chrome, you can view it by [opening Developer Tools](https://developer.chrome.com/docs/devtools/open/) (usually <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>I</kbd> or on Mac <kbd>Cmd</kbd>+<kbd>Option</kbd>+<kbd>I</kbd>), where it has its own tab named ‘Snowplow’ (look for the tab adjacent to ‘Elements’, ‘Console’, ‘Network’, etc. - it may be necessary to expand the list of tabs using the chevrons button (»)). 
```

Author: I'm a sucker for passive voice. :stuck_out_tongue: I've included all your suggestions.

![A screenshot of the Snowplow Inspector extension listing two events observed on the current page. A page_view event is selected, detailing the properties collected as a part of that event.](images/using-poplin-chrome-extension.png)

Within the parsed out HTTP requests that comprise the event payload are a variety of [Snowplow canonical event](/docs/understanding-your-pipeline/canonical-event/index.md) fields such as unique `event_id`, timestamps, and user and session identifiers, as well as any custom [Event](/docs/understanding-your-pipeline/glossary-of-terms/index.md#self-describing-event) or [Entity](/docs/understanding-your-pipeline/glossary-of-terms/index.md#entity) fields.

## The role of Snowplow Inspector in event QA

The Snowplow Inspector extension is an invaluable tool for anyone who is pushing tracking live as it gives real time - easy to access - information on which events are being emitted and what the contents of these events are.
This makes the tool a good first port of call when trying to answer questions such as “why is X event not appearing in my data warehouse” by showing whether or not the event is actually firing on a chosen webpage.  

> Additionally the extension can be configured to show whether or not an event has passed validation according to any event validation rules codified in the corresponding JSON Schema stored in an [Iglu](/docs/understanding-your-pipeline/glossary-of-terms/index.md#iglu) registry. 

Reviewer: ```suggestion
Additionally, you can configure the extension to show whether or not an event has passed validation according to any event validation rules codified in the corresponding JSON Schema stored in an [Iglu](/docs/understanding-your-pipeline/glossary-of-terms/index.md#iglu) registry. 
```

For events that failed validation in production historically that you are unable to replicate in your own browser, see our guides on [how to query failed events](/docs/managing-data-quality/failed-events/failed-events-in-athena-and-bigquery/index.md) from their respective destinations.
These failed events have a [specific format](/docs/managing-data-quality/failed-events/understanding-failed-events/index.md) that includes an array of helpful, detailed error messages that explain the exact reasons why the event failed validation.
These events can also be imported into the extension to view as if your browser had generated them itself.

> ## Event validation with custom Iglu Registries

Reviewer 2: I would remove "Iglu" from the headline here as it could be confusing (local repos and the DS API doesn't match well with the "Iglu Registry" description)

By default the extension is able to see Iglu schemas for Events and Entities that are generated by the official Snowplow [trackers](/docs/collecting-data/collecting-from-own-applications/index.md), as they are publicly available on the public [Iglu Central](/docs/pipeline-components-and-applications/iglu/iglu-repositories/iglu-central/index.md) registry.
This registry is configured by default in most Snowplow Pipelines.
By default, the extension will also create a "Local Registry", which lets you iterate on schemas without having to deploy them to a real registry.

> However the extension can be configured to compare events against private schema repositories (that contain your organization's proprietary schemas).

Reviewer: ```suggestion
However, you can configure the extension to compare events against private schema repositories (that contain your organization's proprietary schemas).
```

> If the extension is not configured to point towards your private schema repository then any custom events or entities will not be recognized by the extension and can not be automatically validated.

Reviewer: ```suggestion
If you skip this step, the extension will not recognize any custom events or entities, and will not be able to validate them.
```

Clicking on the “Unrecognised” button will present you with the “Edit Schema Registry” dialog (covered in the next section) to configure your own registry:

![A Self Describing Event called 'example_event' is displayed in the Snowplow Inspector extension. It contains a property 'example_field_1' with the value 'abc'. The extension says the event's schema is Unrecognized. A cursor graphic is overlaid to indicate this label can be clicked.](images/unrecognized-event.png)
![An image of the form used for configuring a custom Iglu Registry in the Snowplow Inspector extension.](images/edit-registry.png)

> Alternatively, your registries and schemas can be managed by using the ‘Manage Schemas’ button in the top toolbar.

Reviewer: ```suggestion
Alternatively, you can manage your registries and schemas via the ‘Manage Schemas’ button in the top toolbar.
```

> ### Iglu registry and schema management

Reviewer 2: I would rename to "Registries and schema management"

![The Snowplow Inspector extension's Manage Schemas view is shown. Two Iglu registries are listed on the right with buttons for management. The main display has a searchable directory of schemas, grouped by vendor, name, format, and version.](images/manage-schemas.png)

Selecting registries in the list will filter the schema listing to show only the schemas they contain.
You can also search for schemas using the field at the top of the directory.
Use the ‘Clear Filters’ button to clear these and see the full list of schemas again.

Using the ‘Registries...’ button on the right, you can:

- Add more custom registries
- Edit existing registries you've previously configured
- Remove configured registries so they aren't accessible to the extension any more
- Import a list of registries from an existing [Iglu Resolver configuration](/docs/pipeline-components-and-applications/iglu/iglu-resolver/index.md)

Selecting a "local" registry will activate the ‘Schemas...’ button and allow you to manage the schemas it contains.

When adding or editing a registry, the following types are supported:
1. [Local](#local-registries)

> 1. [Data Structures API](#data-structures-api-registries)

Reviewer: I know markdown supports this, but imho 1 2 3 4 reads nicer for those looking at the page source

Author: Fair! I think this is a bad habit I got from asciidoc because it lets me swaps lines around faster when editing. Updated.

1. [Iglu Server](#iglu-server-registries)
1. [Static HTTP](#static-registries)

Configuring a registry requires a name and one of these types.
You may also optionally include a priority, and list of vendor prefixes.
Each type of registry may require additional configuration like API keys to function.

Once configured, schemas from the registry should load in the ‘Manage Schemas’ view, and events using those schemas will be checked for validation.

Now that the extension can query your private Iglu repositories the extension will be able to tell you in real time whether your events and entities are failing validation from entirely within the browser.
An example of an event passing validation is shown below:

![A Self Describing Event called 'example_event' is displayed in the Snowplow Inspector extension. It contains a property 'example_field_1' with the value 'abc'. The extension says the event is Valid against its schema, signalled by green text.](images/event-passing-validation.png)

> An example of an event failing validation is shown below, the reason for this event failing validation is due to “example_field_1” being specified as allowing only the "string" type, meaning it cannot be null.

Reviewer: ```suggestion
An example of an event failing validation is shown below, the reason for this event failing validation is that `example_field_1` is specified as allowing only the "string" type, meaning it cannot be null.
```

Hovering over the "Invalid" warning displays text describing which aspect of the payload failed the validation.
Clicking the warning copies this error to your clipboard.

![A Self Describing Event called 'example_event' is displayed in the Snowplow Inspector extension. It contains a property 'example_field_1' with a null value. The extension says the event is Invalid against its schema, signalled by a bright red background.](images/event-validation-failed.png)

> This is useful, but when it comes to the debugging of failed events and the investigation of exactly WHY a particular event has failed validation there are additional tools that should be considered.

Reviewer: ```suggestion
This is useful, but when it comes to the debugging of failed events and the investigation of exactly _why_ a particular event has failed validation there are additional tools that should be considered.
```

Reviewer: What are those additional tools? We leave the reader hanging

Author: Woops; this was left over from the original page, intended to remove it.

These schemas are managed within the extension and stored on your own machine.

The only required configuration is a name, and the schemas themselves.

#### Data Structures API registries
This type is recommended for use with [Snowplow BDP](/docs/getting-started-with-snowplow-bdp/index.md) using the [Data Structures API](/docs/understanding-tracking-design/managing-data-structures-via-the-api-2/index.md).

In order to function, the extension requires:

- Organization ID: This is usually found in the URL when logged into the Snowplow BDP console. See more at [Managing Console API authentication](/docs/using-the-snowplow-console/managing-console-api-authentication/index.md#credentials-ui-v2).
- API Key: When logged into the Snowplow BDP console, should be available in [API keys for managing Snowplow](https://console.snowplowanalytics.com/credentials) (within "Manage organization"). See more at [Managing Console API authentication](/docs/using-the-snowplow-console/managing-console-api-authentication/index.md#credentials-ui-v2).

> Optionally you may specify:

Reviewer 2: This left me scratching my head, do we know that there is any alternative implementation?

Reviewer 2: I'm now thinking this is probably there to test against Next? However this could be confusion to customers and is really only relevant to Snowplow engineers, I would guess

Author: It's mostly here for completion, do you think it would be better off undocumented or with those use-cases explained?

Reviewer 2: I'm inclined to suggest that we remove it, if I were a customer most likely after reading this I'd reach out to support and ask what are the alternative DS API implementations :)

Because there is no API for listing the contained schemas, the extension will usually attempt to request any referenced schema from static registries to see if it includes it.
To address this imbalance with the other registry types, [`igluctl`](/docs/pipeline-components-and-applications/iglu/igluctl-2/index.md) version 8 added support creating and uploading a file listing all the schemas in a registry.
If the extension detects such a "manifest", it will successfully fetch and display schemas for you without events having to reference the schemas first.

To connect to your static registry, the extension will need:
- Base URI: The base URL to use when requesting schemas. The extension will request schemas relative to this base, expecting a `schemas/{vendor}/{name}/{format/{version}` format.

Optionally you can also provide:
- Manifest URI: The URI to use for the schema manifest (aka [Schema List](https://discourse.snowplow.io/t/igluctl-0-8-0-released/5551#2-uploads-schema-lists-by-default-2)) file that lists the schemas contained in the registry. If not provided, defaults to `schemas`, to match the `igluctl` default.

> ## Events from other devices

Reviewer: This feels like “advanced usage”. If we follow the suggestion to promote this to a higher level page, we can split this part out into its own page. (Similar to the Micro guide.)

Author: Again, here for completion's sake. Agreed it's out of place in a tutorial. Maybe wrap it in `<details>` or something until there's a better home for it?
